### PR TITLE
TDI-39277 dynamic schema case insensitive

### PR DESCRIPTION
**Pull request by mbasiuk-talend**
**What is the current behavior?** (You can also link to an open issue here)
Dynamic fileds were not recognized if upper case runtime schema.

**What is the new behavior?**
Make it case insensitive.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [x] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/895/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/895/)

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java [hunk 1]
```diff
@@ -181,23 +188,40 @@ public void write(Object datum) throws IOException {
...
+                    }
+                }
+                if (!isAdded) {
+                    collectedFields.add(null);
                 }
             }
             isFirst = false;
+            if (collectedFields.size() == 0) {
+                throw new ComponentException(new DefaultErrorCode(HttpServletResponse.SC_BAD_REQUEST, "message"),
+                        new ExceptionContextBuilder().put("message", "Disaster!").build());
```
Comment (sgandon) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: Could you be more explicite in the message please.

File: components/components-snowflake/src/test/java/org/talend/components/snowflake/runtime/SnowflakeWriterTest.java [hunk 2]
```diff
@@ -186,6 +187,30 @@ public void testWrite() throws Exception {
 
     }
 
+    @Test
+    public void testWriteDynamicFields() throws IOException {
```
Comment (sgandon) [inline components/components-snowflake/src/test/java/org/talend/components/snowflake/runtime/SnowflakeWriterTest.java]: I don't see here how you test the case insensitive issue here can you explain to me how you test it ?

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java (hunk 1 above)
Comment (mbasiuk-talend) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: Sure, it was a raw message for local usage, and by mistake I commited it. I have a better implementation, that will be pushed soon. Thanks!

File: components/components-snowflake/src/test/java/org/talend/components/snowflake/runtime/SnowflakeWriterTest.java (hunk 2 above)
Comment (mbasiuk-talend) [inline components/components-snowflake/src/test/java/org/talend/components/snowflake/runtime/SnowflakeWriterTest.java]: Originally this test was for comparing more incoming columns than snowflake runtime schema had. I agree, need to add lower case and upper case scenario in tests. Thanks.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/898/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/898/)

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java [hunk 3]
```diff
@@ -181,23 +188,40 @@ public void write(Object datum) throws IOException {
...
+            }
         }
 
         for (int i = 0; i < row.length; i++) {
```
Comment (sgandon) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: Could you make this another method please ?

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java [hunk 4]
```diff
@@ -180,24 +198,59 @@ public void write(Object datum) throws IOException {
                     .createIndexedRecordConverter(datum.getClass());
         }
         IndexedRecord input = factory.convertToAvro(datum);
-        List<Schema.Field> fields = input.getSchema().getFields();
+        List<Schema.Field> runtimeFields = mainSchema.getFields();
```
Comment (sgandon) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: not sure to understand the *runtime* prefix for the name of this variable, do you mean this is the remote table fields ? may be *remoteTableFields* is more clear ?

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java [hunk 5]
```diff
@@ -180,24 +198,59 @@ public void write(Object datum) throws IOException {
...
 
-        // input and mainSchema synchronization. Such situation is useful in case of Dynamic
+        /*
```
Comment (sgandon) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: After talking to Pierre, I think this part of the method where you compute the field names to be used in the output schema could be externalized to be shared with other components that require a non case sensitive match. Could you mode this piece of code in the components.common module please ?

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java (hunk 4 above)
Comment (mbasiuk-talend) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeWriter.java]: Yes, your're right. Thanks!

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/923/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/923/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/924/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/924/)

File: components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeAvroRegistry.java [hunk 6]
```diff
@@ -99,6 +100,11 @@ protected Field sqlType2Avro(int size, int scale, int dbtype, boolean nullable,
         return field;
     }
 
+    protected Field wrap(boolean nullable, Schema base, String name, Object defaultValue) {
```
Comment (dmytro-sylaiev) [inline components/components-snowflake/src/main/java/org/talend/components/snowflake/runtime/SnowflakeAvroRegistry.java]: Could you please rename boolean parameter to isNullable?

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/932/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/932/)

File: core/components-common/src/main/java/org/talend/components/common/runtime/DynamicSchemaUtils.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    // Util's class shouldn't have public constructor.
+    private DynamicSchemaUtils() {
+    };
+
+    /**
+     * Compare remote schema and input schema with case insensitive condition for field names.
+     * This method performs the same action on schema fields as LEFT OUTER JOIN in SQL.<br/>
+     * Examples with possible cases:<br/>
+     * <ol>
+     * <li>When <b>remote schema</b> is <b>null</b> or <b>empty</b> - throw an exception with detailed message that you should specify your remote schema correctly.</li>
```
Comment (sgandon) [inline core/components-common/src/main/java/org/talend/components/common/runtime/DynamicSchemaUtils.java]: please specify that it is a ComponentException that is thrown here.

File: core/components-common/src/main/java/org/talend/components/common/runtime/DynamicSchemaUtils.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+     * <li>When <b>input schema</b> is <b>null</b> - throw an exception with detailed message that you should specify your input schema correctly.</li>
+     * <li>When <b>schemas</b> are <b>equal</b> - return all fields.</li>
+     * <li>When <b>remote schema</b> has <b>more</b> fields than input schema - copy add all common fields and fill with nulls other cells in result List to conform remote schema fields size</li>
+     * <li>When <b>remote schema</b> has <b>less</b> fields than input schema - </li>
```
Comment (sgandon) [inline core/components-common/src/main/java/org/talend/components/common/runtime/DynamicSchemaUtils.java]: what happens in that case, it seems the javadoc is missing here!

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/935/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/935/)

**Pull request merged**